[pvr] third blind attempt to fix trac #15942

Jalle19 opened the pull request:
ping @ksooo 

Basically this explicitly checks that the iterator is valid and that the pointer inside it is valid. I have no idea if it actually fixes anything or not.

Jalle19 committed:
[pvr] refactor the way information about the
currently active PVR client is determined in
CPVRGUIInfo. Instead of querying the current
client directly we ask CPVRClients to give us
a summary (SBackend) of each connected client.
These summaries are safe to loop over no matter
what happens to the actual clients in the mean
time. Fixes trac #15942

fritsch commented:
@Paxxi ping - as you fixed PVR iterator abuse last time.

stefansaraev commented:
sorry for the noise, but @opdenkamp should be cc'd too, as he broke it while refuctoring.

Paxxi commented:
So your previous attempt didn't work @Jalle19 ? The changes look perfectly fine at first glance but a more thorough investigation of the pvr code might be needed as we still get this crash.

Not sure if/when I have time to spend on it as win32 is apparently crashing. @opdenkamp might be better suited for it.

Jalle19 commented:
@Paxxi it wasn't me who did the previous two attempts. According to the trac tickets all attempts so far have failed.

Paxxi commented:
Oh, I'm just confused then but anyway

ksooo commented:
It was me doing the first two attempts. Unfortunately, neither attempt did the trick. I had the crash again yesterday - on Linux x86_64.

Regarding the fix: According to @opdenkamp fixing it like you suggest by explicitely ensuring not to deref <code>activeClients.cend()</code> and checking <code>activeClient->second</code> not being NULL in addition (which actually also was eaxtly my very first fix attempt) is not good as it hides the actual root cause of the problem - whatever this may be.

ksooo commented:
... however, I'm totally open to include this as is for Isengard as it actually makes sure that the particular crash in GetDriveSpace due to CPVRClient this pointer being NULL cannot occur any longer.

MilhouseVH commented on xbmc/pvr/PVRGUIInfo.cpp:
```diff
@@ -702,31 +702,29 @@ void CPVRGUIInfo::UpdateBackendCache(void)
-      m_iAddonInfoToggleCurrent = 0;
-    }
+    std::advance(activeClient, std::min(m_iAddonInfoToggleCurrent, activeClients.size()));
   }
```
Building for x86 (Ubuntu 15.04) is failing with this PR:

```
PVRGUIInfo.cpp: In member function ‘void PVR::CPVRGUIInfo::UpdateBackendCache()’:
PVRGUIInfo.cpp:717:88: error: no matching function for call to ‘min(unsigned int&, std::map<int, std::shared_ptr<PVR::CPVRClient> >::size_type)’
     std::advance(activeClient, std::min(m_iAddonInfoToggleCurrent, activeClients.size()));
                                                                                        ^
PVRGUIInfo.cpp:717:88: note: candidates are:
In file included from /usr/include/c++/4.9/bits/char_traits.h:39:0,
                 from /usr/include/c++/4.9/string:40,
                 from /home/neil/projects/kodi.git/xbmc/guilib/GUIMessage.h:333,
                 from /home/neil/projects/kodi.git/xbmc/guilib/IMsgTargetCallback.h:31,
                 from /home/neil/projects/kodi.git/xbmc/Application.h:26,
                 from PVRGUIInfo.cpp:21:
/usr/include/c++/4.9/bits/stl_algobase.h:194:5: note: template<class _Tp> const _Tp& std::min(const _Tp&, const _Tp&)
     min(const _Tp& __a, const _Tp& __b)
     ^
/usr/include/c++/4.9/bits/stl_algobase.h:194:5: note:   template argument deduction/substitution failed:
PVRGUIInfo.cpp:717:88: note:   deduced conflicting types for parameter ‘const _Tp’ (‘unsigned int’ and ‘std::map<int, std::shared_ptr<PVR::CPVRClient> >::size_type {aka long unsigned int}’)
     std::advance(activeClient, std::min(m_iAddonInfoToggleCurrent, activeClients.size()));
                                                                                        ^
In file included from /usr/include/c++/4.9/bits/char_traits.h:39:0,
                 from /usr/include/c++/4.9/string:40,
                 from /home/neil/projects/kodi.git/xbmc/guilib/GUIMessage.h:333,
                 from /home/neil/projects/kodi.git/xbmc/guilib/IMsgTargetCallback.h:31,
                 from /home/neil/projects/kodi.git/xbmc/Application.h:26,
                 from PVRGUIInfo.cpp:21:
/usr/include/c++/4.9/bits/stl_algobase.h:240:5: note: template<class _Tp, class _Compare> const _Tp& std::min(const _Tp&, const _Tp&, _Compare)
     min(const _Tp& __a, const _Tp& __b, _Compare __comp)
     ^
/usr/include/c++/4.9/bits/stl_algobase.h:240:5: note:   template argument deduction/substitution failed:
PVRGUIInfo.cpp:717:88: note:   deduced conflicting types for parameter ‘const _Tp’ (‘unsigned int’ and ‘std::map<int, std::shared_ptr<PVR::CPVRClient> >::size_type {aka long unsigned int}’)
     std::advance(activeClient, std::min(m_iAddonInfoToggleCurrent, activeClients.size()));
                                                                                        ^
In file included from /usr/include/c++/4.9/algorithm:62:0,
                 from /home/neil/projects/kodi.git/xbmc/guilib/Geometry.h:31,
                 from /home/neil/projects/kodi.git/xbmc/cores/IPlayer.h:26,
                 from /home/neil/projects/kodi.git/xbmc/GUIInfoManager.h:37,
                 from PVRGUIInfo.cpp:25:
/usr/include/c++/4.9/bits/stl_algo.h:3439:5: note: template<class _Tp> _Tp std::min(std::initializer_list<_Tp>)
     min(initializer_list<_Tp> __l)
     ^
/usr/include/c++/4.9/bits/stl_algo.h:3439:5: note:   template argument deduction/substitution failed:
PVRGUIInfo.cpp:717:88: note:   mismatched types ‘std::initializer_list<_Tp>’ and ‘unsigned int’
     std::advance(activeClient, std::min(m_iAddonInfoToggleCurrent, activeClients.size()));
                                                                                        ^
In file included from /usr/include/c++/4.9/algorithm:62:0,
                 from /home/neil/projects/kodi.git/xbmc/guilib/Geometry.h:31,
                 from /home/neil/projects/kodi.git/xbmc/cores/IPlayer.h:26,
                 from /home/neil/projects/kodi.git/xbmc/GUIInfoManager.h:37,
                 from PVRGUIInfo.cpp:25:
/usr/include/c++/4.9/bits/stl_algo.h:3444:5: note: template<class _Tp, class _Compare> _Tp std::min(std::initializer_list<_Tp>, _Compare)
     min(initializer_list<_Tp> __l, _Compare __comp)
     ^
/usr/include/c++/4.9/bits/stl_algo.h:3444:5: note:   template argument deduction/substitution failed:
PVRGUIInfo.cpp:717:88: note:   mismatched types ‘std::initializer_list<_Tp>’ and ‘unsigned int’
     std::advance(activeClient, std::min(m_iAddonInfoToggleCurrent, activeClients.size()));
                                                                                        ^
../../Makefile.include:93: recipe for target 'PVRGUIInfo.o' failed
make[1]: *** [PVRGUIInfo.o] Error 1
Makefile:470: recipe for target 'xbmc/pvr/pvr.a' failed
make: *** [xbmc/pvr/pvr.a] Error 2
make: *** Waiting for unfinished jobs....
```

opdenkamp commented:
sorry, had some issues with email not being delivered today that I just fixed.
@stefansaraev didn't say it was me, but it may very well be, don't remember

the issue is that originally the pointers to the guiinfo stuff got updated from the same thread that registered new add-ons and deleted them, and now it gets done from another thread.

the proper solution would be to either do it like before, or iterating over the list of add-ons for guiinfo in pvrclients.cpp instead of in a copy, and returning a shared ptr to the active one to guiinfo. that way you don't have to do bookkeeping in 2 classes and keep the two classes synced.

Jalle19 commented:
jenkins build this please

ksooo commented:
FYI, I had that damn crash again yesterday, this time even while watching a movie (this PR was not applied then, don't get me wrong, please!).

Thus, definitely a +1 from me for this PR in Isengard.

Jalle19 commented:
@ksooo you don't happen to have a crash log with debugging enabled?

ksooo commented:
http://paste.ubuntu.com/11734571/

<pre>
Program terminated with signal SIGSEGV, Segmentation fault.
#0  PVR::CPVRClient::GetDriveSpace (this=0x0, iTotal=iTotal@entry=0x7f922fffed90, iUsed=iUsed@entry=0x7f922fffed88) at PVRClient.cpp:562
</pre>

=> this=0x0 !!!

Jalle19 commented:
Okay so I've pushed a completely new approach. I've even tested it. Basically, instead of calling `GetConnectedClients` from `CPVRGUIInfo` and then looping over the results, we ask `CPVRClients` to give us a summary of all connected clients (`SBackendProperties`). This we can loop over safely.

In addition to that change, I removed the WTF that was `AddonInfoToggle()`, now the "current active client" is updated on every call to `UpdateBackendCache()`. The main thread loop was modified to call this function at the defined "toggle interval" instead of every 10 seconds.

If this doesn't work then we're all doomed.

opdenkamp commented:
That's what I suggested, that should fix it.

And as explained before, it wasn't a WTF in the way it was used in the 
old code, but it broke with the current code.


On 18-06-15 12:39, Sam Stenvall wrote:
> Okay so I've pushed a completely new approach. I've even tested it.
> Basically, instead of calling |GetConnectedClients| from |CPVRGUIInfo|
> and then looping over the results, we ask |CPVRClients| to give us a
> summary of all connected clients (|SBackendProperties|). This we can
> loop over safely.
>
> In addition to that change, I removed the WTF that was
> |AddonInfoToggle()|, now the "current active client" is updated on every
> call to |UpdateBackendCache()|. The main thread loop was modified to
> call this function at the defined "toggle interval" instead of every 10
> seconds.
>
> If this doesn't work then we're all doomed.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/xbmc/xbmc/pull/7291#issuecomment-113103996>.
>

Jalle19 commented:
No disrespect intended, I just found it a bit over-complicated.

Jalle19 commented:
jenkins build this please

ksooo commented on xbmc/pvr/PVRGUIInfo.cpp:
```diff
@@ -691,92 +670,44 @@ void CPVRGUIInfo::CharInfoProvider(std::string &strValue) const
-  m_iBackendTotalDiskspace      = iBackendkBTotal;
+  m_strBackendName = properties.backendName;
+  m_strBackendVersion = properties.backendVersion;
+  m_strBackendHost = properties.backendHost;
```
Well... what about disk space? Not handled here any longer.

Jalle19 replied:
Crap, I totally missed that. Won't have time to code until Sunday.

Jalle19 commented:
jenkins build this please

Jalle19 commented:
@ksooo I added the missing code for handling the disk space.

IMO this is for Isengard, what do you guys think?

Jalle19 commented:
@ksooo please review the latest commit, it fixes the issue you brought up about needlessly updating the backend information.

ksooo commented:
@Jalle19 sorry, caching the props does not make things better, see my latest code comments.

ksooo commented:
@Jalle19 okay, caching makes things better., But not good on the sense of actually working :-) 

What makes me nervous is that the code gets more and more complex now, but the underlying problem still remains the same. If you use a numeric value to remember the last client, this index might get to large if the number of connected clients decreases between two calls to UpdateBackendCache. Thus, you need to check the index before accessing the vector/array. This is the essential part of c the fix, this preventa the crash. In previous fix attempts this was done by checking the size of the vector of the connected clients. Simple and working. Now, this is done via checking the count of connected clients as returned in the array. Instead of just changing two lines of code ti prevent the crash, we now change much more code, copying data, do caching, etc but still do not fix the actual problem.

Jalle19 commented:
> What makes me nervous is that the code gets more and more complex now, but the underlying problem still remains the same

I wouldn't call this complex.

> this index might get to large if the number of connected clients decreases between two calls to UpdateBackendCache

This doesn't matter anymore. The same `SBackendInformation` (including the number of connected clients) is used until the "current index" wraps back to zero, in which case the backend information is updated. You could add/delete PVR clients a million times in between the calls to `UpdateBackendCache` and it wouldn't matter since `CPVRGUIInfo` doesn't operate directly on anything from `CPVRClients` anymore.

> Now, this is done via checking the count of connected clients as returned in the array. Instead of just changing two lines of code ti prevent the crash, we now change much more code, copying data, do caching, etc but still do not fix the actual problem

Sure this is more code, but I thought we all agreed that just adding a null check wasn't a real solution?

ksooo commented:
Sorry, no offense intended, but I still think that the changes do not actually improve the situation compared with just checking the size of the container with the currently connected clients. It's basically the same approach, not safer not more efficient, but with more code complexity (then before)  - copying of data, syncing cached data. For what reason? How does this improve things compared to the no-brainer solution? Sorry, trying hard, but don't get it.

Jalle19 commented:
If everyone else is okay with the simple null check patch then we can go with that as well.

ksooo commented:
Citing @opdenkamp : "the proper solution would be [...] iterating over the list of add-ons for guiinfo in pvrclients.cpp instead of in a copy, and returning a shared ptr to the active one to guiinfo.

Jalle19 commented:
I don't see how that would be better than checking for null.

ksooo commented:
Well, citing @opdenkamp again "that way you don't have to do bookkeeping in 2 classes and keep the two classes synced." ... This is a cleaner design, imo... and it actually works in all cases. The class that modifies the list of clients takes care of returning only valid clients to its users. Using class needs not to do any validity checks for client is gets from pvrclients and there is no need anymore for the (as a wrote earlier not always working numeric index stuff in guiinfo)...

Jalle19 commented:
I don't see how returning a shared pointer to an addon will solve this particular issue since we already end up in a situation where we have to check if the pointer is null. Furthermore, the index in GUIInfo is requires in any case, I don't see how we could get rid of it?

ksooo commented:
I will sketch the code and send you a pointer to it, so you can get the gist of the idea.

ksooo commented:
@Jalle19 here is my take on the fix approach @opdenkamp has mentioned. => https://github.com/ksooo/xbmc/commit/a07a93dc641d9b4720eb9b511f12e007843cc43b

Plesae note that <code>CPVRGUIInfo::m_iActiveClient</code> and <code>CPVRGUIInfo::m_iActiveClients</code> are now only needed to implement <code>CPVRGUIInfo::CharInfoBackendNumber</code>. This this pure UI stuff; no real logic acts upon these members anymore. All logic that manages available clients and iterating them is now in <code>CPVRClients</code>.

IMO, the code is more clean now and the crash cannot happen any longer. What do you think?

opdenkamp commented:
+100

Jalle19 commented:
I'm not sure I like it. To me your approach is more complicated than mine, plus now `CPVRClients` has to keep track of which client is "active" when the only piece of code in the entire application that needs this is `CPVRGUIInfo`. The new member variables even have "guiInfo" in their names.

opdenkamp commented:
Self-containment principle... The list of add-ons is maintained by cpvraddons and can become invalid over time. You can try to maintain a pointer in a list that's changed by something external to the class from which you're reading it, and try to patch the issues that you'll run into, or you can maintain the pointer in the class that changes the list.

Re. name: change it to something else if you don't like guiinfo in it ;-)

ksooo commented:
Idrc about the names of the variables. Come up with something more appropriate and I will be happy to change them.

Jalle19 commented:
@opdenkamp I solved that by returning a vector that is not linked in any way to the client map, except it generally has the same number of elements. `CPVRClients` shouldn't have to care about what `CPVRGUIInfo` does/needs (like keeping track of the "active" client), and `CPVRGUIInfo` shouldn't have to bother with what goes on/could go on in `CPVRClients` (like a client being erased from the map). My approach is IMO the only one so far that has avoided both of these issues.

@ksooo the names as such are fine

ksooo commented:
@Jalle19 unrelated to the current discussion about the best approch, I just stumbled over one thing in your code. You should eliminate duplicate data. All the stuff in <code>m_strBackendName</code> ff is also present in <code>m_backendProperties</code>. All these members should be obsolete now.

Jalle19 commented:
@ksooo http://trac.kodi.tv/ticket/15942#comment:22 initial reports are good. I suggest I fix the duplicate data issue you pointed out then we merge this.

opdenkamp commented:
the behaviour of the iteration is pretty random when you do it like this, instead of how @ksooo and I think it should be done... Not that it's very important stuff, as it's only gui info.

Also, you've added `struct SBackend`, which is just duping data and is not needed if you just return a shared ptr to the add-on.

-1 like this.

ksooo commented:
I consider none of the two alternatives the Holy Grail. Both approaches have pros and cons. I tend to prefer my approach but will not block @Jalle19's take.

opdenkamp commented:
that new struct is pointless.

ksooo commented:
@opdenkamp side question: What is this <code>CPVRClient:IsReadyToUse</code> stuff about? Can the ready-state change at any time? If yes, then every access to CPVRClient instance' mamber functions must be protected against sudden "non-readyness"?

opdenkamp commented:
yes, e.g. the connection can drop.

ksooo commented:
@opdenkamp @Jalle19 time is running... Do we want to do anything about this issue for Isengard? IMO, if we do not merge this PR (or decide to go "my" way) now, we should take at least the no-brainer null check workaround for Isengard to preevnt that nerving random crashes.

Jalle19 commented:
I'll push something soon to address the duplication of data.

xhaggi commented on xbmc/pvr/addons/PVRClients.h:
```diff
@@ -45,12 +45,35 @@ namespace PVR
+   */
+  struct SBackendProperties
+  {
+    unsigned int numBackends = 0;
```
what's the reason for using this special struct and not `backends.size()`?

ksooo replied:
Good catch.

Jalle19 replied:
Good catch indeed, will change it.

ksooo commented:
@Jalle19, @opdenkamp  Unfotunately we still don't have agreement on the right approach. As I already said, I could live with both alternatives. You guys should come to an agreement soon, as either of you actually gave -1 to the "other" approach.

Jalle19 commented:
If you decide on this approach then please let me know so I can squash it down.

jenkins build this please

ksooo commented:
@jalle19 c'mon! You are part of the decision making process. ;-)

Jalle19 commented:
@ksooo sure, but I'm out of paint.

xhaggi commented on xbmc/pvr/addons/PVRClients.h:
```diff
@@ -45,12 +45,28 @@ namespace PVR
+    long long   diskTotal = 0;
+  };
+
+  typedef std::vector<SBackend> SBackendProperties;
```
we should drop the prefix `S` because it's not a struct anymore

Jalle19 replied:
You're right. I don't think I have time to code anymore today though.

opdenkamp commented:
+1for this aproach

ksooo commented:
@Jalle19 Do you plan to merge this for RC1?

Jalle19 commented:
Yes, give me a few hours.

ksooo commented:
@Jalle19 cool. Thx.

Jalle19 commented:
@xhaggi I dropped the `typedef` completely, it wasn't really necessary anyway. Also squashed to one commit with a slightly better description of the change.

jenkins build this please

ksooo commented:
Looking good.

Jalle19 commented:
build failure unrelated (addons)

xhaggi commented:
@Jalle19 thanks looks good

Jalle19 commented:
so, okay to merge?

xhaggi commented on xbmc/pvr/addons/PVRClients.h:
```diff
@@ -45,12 +45,26 @@ namespace PVR
+  {
+    std::string backendName;
+    std::string backendVersion;
+    std::string backendHost;
```
minor but i would drop backend as a prefix.

xhaggi commented on xbmc/pvr/addons/PVRClients.cpp:
```diff
@@ -258,6 +258,40 @@ bool CPVRClients::GetClientName(int iClientId, std::string &strName) const
+    if (!client->ReadyToUse())
+      continue;
+
+    SBackend properties;
```
we should name the var like the struct `backend`

Jalle19 replied:
I disagree with this since the struct holds information about a backend ("properties"), it's not a backend on its own. Same goes for ´m_backends`.

xhaggi replied:
so why do you call it `SBackend` that's confusing?

Jalle19 replied:
Basically because SBackendProperties would be one object, but what would you call a vector of those? You can't write `properties'` like in standard English.

xhaggi replied:
sure you could only use `m_backendsProperties` or add a prefix/suffix to indicate that's a vec/list.

Jalle19 replied:
I see your point, but TBH I don't think this is worth arguing over any longer.

xhaggi replied:
yep :)

opdenkamp replied:
and furthermore, using not using `auto` where it's not actually replacing something like `std::map<foo, struct bar>::const_iterator` would make things less confusing for readers too (comment doesn't apply to this instance, but some people sure like using the `auto` keyword as much as possible ;-))

xhaggi commented on xbmc/pvr/PVRGUIInfo.h:
```diff
@@ -179,13 +176,11 @@ namespace PVR
     XbmcThreads::EndTime            m_ToggleShowInfo;
     EPG::CEpgInfoTagPtr             m_playingEpgTag;
-
+    std::vector<SBackend>           m_backendProperties;
```
i would name the member `m_backends`

xhaggi commented on xbmc/pvr/addons/PVRClients.h:
```diff
@@ -147,6 +161,12 @@ namespace PVR
+     * @brief Returns properties about all connected clients
+     * @return the properties
+     */
+    std::vector<SBackend> GetBackendProperties() const;
```
GetBackends

xhaggi commented:
only some minors about method/var naming, ignore it if you disagree, but imo it's more reasonable

Jalle19 commented:
Sure, it's a leftover from all the work in progress. I'll squash a fix tomorrow morning and merge if no one has any further objections.

ksooo commented on xbmc/pvr/PVRGUIInfo.cpp:
```diff
@@ -452,8 +424,10 @@ int CPVRGUIInfo::TranslateIntInfo(DWORD dwInfo) const
   {
-    if (m_iBackendTotalDiskspace > 0)
-      iReturn = (int) (100 * m_iBackendUsedDiskspace / m_iBackendTotalDiskspace);
+    auto &backend = GetCurrentActiveBackend();
```
Nitpick 1: maybe const auto...?

ksooo commented on xbmc/pvr/PVRGUIInfo.cpp:
```diff
@@ -452,8 +424,10 @@ int CPVRGUIInfo::TranslateIntInfo(DWORD dwInfo) const
+    auto &backend = GetCurrentActiveBackend();
+
+    if (backend.diskTotal > 0)
+      iReturn = (int)(100 * backend.diskUsed / backend.diskTotal);
```
Nitpick 2: Maybe a C++ cast is better here...

ksooo commented on xbmc/pvr/PVRGUIInfo.cpp:
```diff
@@ -691,92 +677,43 @@ void CPVRGUIInfo::CharInfoProvider(std::string &strValue) const
-      iBackendkBTotal = 0;
-    }
+  // Get the properties for the currently active backend
+  auto backend = GetCurrentActiveBackend();
```
const auto?

Jalle19 commented:
jenkins build and merge

Jalle19 commented:
Jenkins can't merge anymore?

MartijnKaijser commented:
There was a build failure on rpi with Addons so it wouldn't auto merge

xhaggi commented on xbmc/pvr/addons/PVRClients.cpp:
```diff
@@ -258,6 +258,40 @@ bool CPVRClients::GetClientName(int iClientId, std::string &strName) const
+      continue;
+
+    SBackend properties;
+
```
+1 as it also makes it more review-fiendly
